File: 889_construct_binary_tree_from_preorder_and_postorder_traversal.cpp
class Solution {
public:
    TreeNode* bz(vector<int>& p, vector<int>& q) {
        if (p.empty()) return nullptr;
        TreeNode* n = new TreeNode(p[0]);
        if (p.size() == 1) return n;
        int d = find(q.begin(), q.end(), p[1]) - q.begin() + 1;
        vector<int> a(p.begin() + 1, p.begin() + d + 1), b(q.begin(), q.begin() + d);
        vector<int> c(p.begin() + d + 1, p.end()), e(q.begin() + d, q.end() - 1);
        n->left = bz(a, b);
        n->right = bz(c, e);
        return n;
    }
    TreeNode* constructFromPrePost(vector<int>& p, vector<int>& q) {
        return bz(p, q);
    }
};

File: 889_construct_binary_tree_from_preorder_and_postorder_traversal.py
class Solution:
    def constructFromPrePost(self, p, q):
        if not p: return None
        n = TreeNode(p[0])
        if len(p) == 1: return n
        d = q.index(p[1])
        n.left = self.constructFromPrePost(p[1:d+2], q[:d+1])
        n.right = self.constructFromPrePost(p[d+2:], q[d+1:-1])
        return n
